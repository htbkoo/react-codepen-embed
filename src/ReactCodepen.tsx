import * as React from 'react';
import {Component, ComponentClass, FunctionComponent} from 'react';
import CodepenEmbedScriptTagBuilder, {ScriptTagBuilder} from "./CodepenEmbedScriptTagBuilder";

type ErrorType = string;

interface LoaderProps {
    isLoading: boolean,
    error: ErrorType
}

interface ReactCodepenProps {
    defaultTab: string,
    hash: string,
    height: number,
    loader?: FunctionComponent<LoaderProps> | ComponentClass<LoaderProps> | string,
    preview: boolean,
    title?: string,
    themeId: string | number,
    user: string,
    version: number,
    shouldLoadScript: boolean,
    overrideAsLoaded?: boolean
}

interface ReactCodepenState {
    loaded: boolean,
    loading: boolean,
    error?: ErrorType,
}

class ReactCodepen extends Component<ReactCodepenProps, ReactCodepenState> {
    private _mounted: boolean;
    private scriptTagBuilder: ScriptTagBuilder;

    static defaultProps = {
        defaultTab: 'css,result',
        height: 300,
        preview: true,
        themeId: 'dark',
        version: 2,
        shouldLoadScript: true
    };

    state = {loaded: false, loading: true, error: undefined};

    constructor(props: ReactCodepenProps, context) {
        super(props, context);

        this.scriptTagBuilder = new CodepenEmbedScriptTagBuilder()
            .setAsync(true)
            .withOnLoadHandler(this.handleScriptLoad.bind(this))
            .withOnErrorHandler(this.handleScriptError.bind(this));
    }

    componentDidMount() {
        this._mounted = true;

        if (this.props.shouldLoadScript) {
            // load the codepen embed script
            this.scriptTagBuilder.appendTo(document.body);
        }
    }

    componentWillUnmount() {
        this._mounted = false;
    }

    render() {
        if (!this.isLoaded() && this.props.loader) {
            const loaderComponent = React.createElement(this.props.loader, {
                isLoading: this.state.loading,
                error: this.state.error
            });
            return (
                <div
                    data-height={this.props.height}
                    data-theme-id={this.props.themeId}
                    data-slug-hash={this.props.hash}
                    data-default-tab={this.props.defaultTab}
                    data-user={this.props.user}
                    data-embed-version={this.props.version}
                    data-pen-title={this.props.title}
                    data-preview={this.props.preview}
                    className="codepen"
                >
                    {loaderComponent}
                </div>
            );
        } else if (this.isLoaded()) {
            const penLink = `https://codepen.io/${this.props.user}/pen/${this.props.hash}/`;
            const userProfileLink = `https://codepen.io/${this.props.user}`;

            return (
                <div
                    data-height={this.props.height}
                    data-theme-id={this.props.themeId}
                    data-slug-hash={this.props.hash}
                    data-default-tab={this.props.defaultTab}
                    data-user={this.props.user}
                    data-embed-version={this.props.version}
                    data-pen-title={this.props.title}
                    data-preview={this.props.preview}
                    className="codepen"
                >
                    See the Pen <a href={penLink}>{this.props.title}</a>
                    by {this.props.user} (<a href={userProfileLink}>@{this.props.user}</a>)
                    on <a href="https://codepen.io">CodePen</a>.
                </div>
            );
        } else {
            return null;
        }
    }

    private isLoaded() {
        return this.props.overrideAsLoaded || this.state.loaded;
    }

    private handleScriptLoad() {
        // do not do anything if the component is already unmounted.
        if (!this._mounted) return;

        this.setState({
            loaded: true,
            loading: false
        });
    }

    private handleScriptError() {
        if (!this._mounted) return;

        this.setState({
            error: 'Failed to load the pen'
        });
    }
}

export default ReactCodepen;