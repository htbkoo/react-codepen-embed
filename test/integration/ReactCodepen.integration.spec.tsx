import * as React from "react";
import {when} from 'jest-when'
import {mount} from "enzyme";

import ReactCodepen from "../../src";

describe('<ReactCodepen/> - integration tests', function () {
    it('should mount without error', function () {
        // given
        mount(<ReactCodepen hash="someHash" user="someUser" shouldLoadScript={false}/>);
        // when
        // then
    });
});