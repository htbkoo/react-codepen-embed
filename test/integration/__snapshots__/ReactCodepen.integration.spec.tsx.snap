// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReactCodepen/> - integration tests Snapshot tests Not loading script should, when loaded, render and match the snapshot 1`] = `
<div
  className="codepen"
  data-default-tab="defaultTab"
  data-embed-version={100}
  data-height={512}
  data-pen-title="title"
  data-preview={false}
  data-slug-hash="hash"
  data-theme-id="theme"
  data-user="user"
>
  <div>
    See the Pen 
    <a
      href="https://codepen.io/user/pen/hash/"
    >
      title
    </a>
    by 
    user
     (
    <a
      href="https://codepen.io/user"
    >
      @
      user
    </a>
    ) on 
    <a
      href="https://codepen.io"
    >
      CodePen
    </a>
    .
  </div>
</div>
`;

exports[`<ReactCodepen/> - integration tests Snapshot tests Not loading script should, when loader is provided and not loaded, render and match the snapshot 1`] = `
<div
  className="codepen"
  data-default-tab="css,result"
  data-embed-version={2}
  data-height={300}
  data-preview={true}
  data-slug-hash="someHash"
  data-theme-id="dark"
  data-user="someUser"
>
  <p>
    loader
  </p>
</div>
`;

exports[`<ReactCodepen/> - integration tests Snapshot tests Not loading script should, when not loaded and no loader provided, render and match the snapshot 1`] = `
<div
  className="codepen"
  data-default-tab="css,result"
  data-embed-version={2}
  data-height={300}
  data-preview={true}
  data-slug-hash="someHash"
  data-theme-id="dark"
  data-user="someUser"
/>
`;
